ONSdigital/eq-author-app
# Ear 894 develop back end for folders

> BEFORE MAKING YOUR PR... There must be no linting errors and all tests must pass

### What is the context of this PR?

The big ol' back-end re-write to incorporate `folders`

Note: if trying to output to the console within any of the tests, remove the `--forceExit` argument in the test script file under `scripts/`.

Also includes:

- front-end updates so Author works with modified schema
- updates to publisher v2 so it works as-is with folders in the schema (will need further work when we want to add skip conditions to folders)

### How to review 

This is gonna take a fairly long time to review and might need a all hands on approach. It will help people become familiar with the back-end (I wasn't that well) and how the data is structured going forward.

Testing of this makes heavy use of the Apollo dev tools.

Hitting the explorer button in the Apollo dev tools opens up all the mutations and queries we have. Most of the operations will be under mutations.
I attempted to test each one of these to see if they perform the action against an exported schema.

Get exported schema from:
`http://localhost:4000/export/<questionnaire-id-goes-here>`

So for example using `createAnswer`

- Add the relevant `questionPageId` and whatever answer type you want.
- Hit the play button and the result should appear on the right. Depending if you added any return values.
- Refresh the export page and check the answer is there
- Tada!
